src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -311,22 +322,40 @@ private static string GenerateUpgradeReport(string projectName, string backupPat
             packageEnumerationTime.Start();
             try
             {
-                // collect the install state of the existing packages
-                foreach (IProjectContextInfo project in uiService.Projects)
+                IServiceBroker sb = uiService.UIContext.ServiceBroker;
+                if (!userAction.IsSolutionLevel && userAction.Action == NuGetProjectActionType.Install && uiService.Projects?.Count() == 1)
                 {
-                    IEnumerable<IPackageReferenceContextInfo> installedPackages = await project.GetInstalledPackagesAsync(
-                        uiService.UIContext.ServiceBroker,
-                        cancellationToken);
-
-                    foreach (IPackageReferenceContextInfo package in installedPackages)
+                    // expected only 1 PackageReference project
+                    IProjectContextInfo prj = uiService.Projects.First();
+                    if (prj.ProjectStyle == ProjectModel.ProjectStyle.PackageReference && prj.ProjectKind == NuGetProjectKind.PackageReference)
                     {
-                        Tuple<string, string> packageInfo = new Tuple<string, string>(
-                            package.Identity.Id,
-                            (package.Identity.Version == null ? "" : package.Identity.Version.ToNormalizedString()));
+                        IInstalledAndTransitivePackages installedAndTransitives = await prj.GetInstalledAndTransitivePackagesAsync(sb, cancellationToken);
+                        foreach (IPackageReferenceContextInfo package in installedAndTransitives.InstalledPackages)
+                        {
+                            AddToExisting(package, existingPackages);
+                        }
 
-                        if (!existingPackages.Contains(packageInfo))
+                        transitivePackageIds = installedAndTransitives.TransitivePackages
+                            .Select(pkg => VSTelemetryServiceUtility.NormalizePackageId(pkg.Identity.Id)).Distinct().ToHashSet();
+                    }
```

Comment:
I believe we don't need `Distinct` here, because `.ToHashSet()` is just `new HashSet(source)` end of the so it should distinct by default.

Reply:
Replaced with for loop.